pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,171 @@
+	nodeLister v1lister.NodeLister
+
+	cloud cloudprovider.Interface
+
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// A map presenting the node and whether it currently exists
+	taggedNodes map[string]bool
+
+	// A map representing the nodes that were ever in the cluster
+	nodeMap map[string]*v1.Node
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration) (*TaggingController, error) {
+
+	tc := &TaggingController{
+		kubeClient:        kubeClient,
+		nodeLister:        nodeInformer.Lister(),
+		cloud:             cloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		taggedNodes:       make(map[string]bool),
+		nodeMap:           make(map[string]*v1.Node),
+	}
+
+	return tc, nil
+}
+
+// Run will start the controller to tag resources attached to a cluster
+// and untag resources detached from a cluster.
+func (tc *TaggingController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	wait.UntilWithContext(ctx, tc.monitorNodes, tc.nodeMonitorPeriod)
+}
+
+func (tc *TaggingController) monitorNodes(ctx context.Context) {
+	nodes, err := tc.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("error listing nodes from cache: %s", err)
+		return
+	}
+
+	for k := range tc.taggedNodes {
+		tc.taggedNodes[k] = false
+	}
+
+	var nodesToTag []*v1.Node
+	for _, node := range nodes {
+		if _, ok := tc.taggedNodes[node.GetName()]; !ok {
+			nodesToTag = append(nodesToTag, node)
+		}
+
+		tc.nodeMap[node.GetName()] = node
+		tc.taggedNodes[node.GetName()] = true
+	}
+	tc.tagNodesResources(nodesToTag)
+
+	var nodesToUntag []*v1.Node
+	for nodeName, existed := range tc.taggedNodes {
+		if existed == false {
+			nodesToUntag = append(nodesToUntag, tc.nodeMap[nodeName])
+		}
+	}
+	tc.untagNodeResources(nodesToUntag)
+
+	tc.syncDeletedNodesToTaggedNodes()
+}
+
+// tagNodesResources tag node resources from a list of node
+// If we want to tag more resources, modify this function appropriately
+func (tc *TaggingController) tagNodesResources(nodes []*v1.Node) {
+	for _, node := range nodes {
+		klog.Infof("Tagging resources for node %s.", node.GetName())
+	}
+}
+
+func (tc *TaggingController) untagNodeResources(nodes []*v1.Node) {
+	for _, node := range nodes {
+		klog.Infof("Untagging resources for node %s.", node.GetName())
+	}
+}
+
+// syncDeletedNodes delete (k, v) from taggedNodes
+// if it doesn't exist
+func (tc *TaggingController) syncDeletedNodesToTaggedNodes() {
+	for k, v := range tc.taggedNodes {
+		if v == false {
+			delete(tc.taggedNodes, k)
+		}
+	}
+}
+
+//// tagEc2Instances applies the provided tags to each EC2 instances in
```

Comment:
Pls ignore this section, I'm in the middle of tagging instances and I don't want to remove it.